### PR TITLE
aria-placeholder tests and results

### PR DESCRIPTION
**Pull request by RavenAlly**
Test cases and results for aria-placeholder.

### CODE REVIEW

**Comment by mfairchild365** [comment]
This is great! Thank you!

I’m kind of picky when it comes to defining expectations, so this feedback isn’t meant to be discouraging, but hopefully constructive. This is a great start!

it looks like you have two expectations defined:

* MUST convey placeholder when unpopulated
* MUST contribute to accessible name

The aria spec uses the language “with no value” instead of “unpopulated”. For consistency, I’m tempted to stick with the “no value” language. What do you think?

It might also be good to document what happens when the field has a value and aria-placeholder. Should screen readers no longer convey the placeholder? Should this be an expectation that we test?

The [accessible name and description computation spec](https://w3c.github.io/accname/) doesn’t factor in placeholder text. So it’s fine if screen readers use the placeholder text as the name if no other name exists, but that isn’t behavior that is “required” by a standard. Unless there is broad consensus across screen readers in how they handle this case, I’d be tempted to make this a “should” or “may” expectation, rather than “must”. What do you think?

**Comment by JAWS-test** [comment]
[HTML AAM](https://w3c.github.io/html-aam/#input-type-text-input-type-password-input-type-search-input-type-tel-input-type-email-input-type-url-and-textarea-element-accessible-name-computation) specifies that placeholder must be used as Accessible Name if no other name exists

**Comment by RavenAlly** [comment]
Thanks @mfairchild365 for the detailed feedback. Obviously I jumped straight in, it might've been worth getting expectations signed off first!

No worries about switching "unpopulated" to "with no value" wording, I appreciate consistency as much as the next person. 👍 

As mentioned by @JAWS-test, I used "MUST contribute to the accessible name" on the basis of [HTML AAM](https://www.w3.org/TR/html-aam-1.0/#input-type-text-input-type-password-input-type-search-input-type-tel-input-type-email-input-type-url-and-textarea-element-accessible-name-computation). I note that [Accessible Name and Description Computation 1.1](https://www.w3.org/TR/accname/) doesn't mention placeholder though. Happy to change to SHOULD or MAY if I'm not using a valid reference.

I agree it would be good to include test results for text fields that contain a value and I'm happy to add that. I wasn't sure how to word the expectation because I couldn't find a reference. I'm thinking "MAY convey aria-placeholder for text field with value". What do you think?

Whilst wading through documentation I also thought I could also add "MUST NOT convey aria-placeholder when (HTML attribute) placeholder is defined" based on [section 3.5 of HTML AAM](https://www.w3.org/TR/html-aam-1.0/#html-attribute-state-and-property-mappings). What do you think?

> When the `placeholder` and `aria-placeholder` attributes are both present, and the `placeholder` attribute's value is non-empty, user agents **_MUST_** expose the value of the `placeholder` attribute, and ignore `aria-placeholder`.

**Comment by mfairchild365** [comment]
I think it’s awesome that you jumped right in! 🙂

The referenced part of HTML AAM discusses how the placeholder attribute might take part in the name computation, but doesn’t mention aria-placeholder (unless I’m missing something). I’m okay with it being a MUST if there is robust support for it, but I think it’s worth considering a lower strength expectation otherwise.

regarding “MAY convey aria-placeholder for text field with value” and the strength for that expectation… it’s not always clear cut. If there is no spec to justify the strength, I usually go with whatever my gut says it should be when drafting the expectation and associated test. Then, once I have all the results, I can review how well supported the expectation is and adjust the strength if my gut didn’t match reality. Does that make sense? 

I like the addition of “MUST NOT convey aria-placeholder when (HTML attribute) placeholder is defined”.

**Comment by RavenAlly** [comment]
I was conflating aria-placeholder with placeholder in HTML AAM because they're listed as equivalents in [section 3.1 of ARIA in HTML](https://www.w3.org/TR/html-aria/#docconformance-attr). My bad. I'll adjust to SHOULD for contributing to accessible name.

I'll be back with additional expectations and results. Cheers!

File: data/tests/tech/aria/aria-placeholder.json [hunk 1]
```diff
@@ -0,0 +1,1693 @@
...
+            "focus_location": "before target"
+          },
+          "after": "target",
+          "output": "\"input left bracket type equals text right bracket aria placeholder defined unlabelled, edit, type in text\"",
+          "results": [
+            {
+              "feature_id": "aria/aria-placeholder_attribute",
+              "feature_assertion_id": "contribute_to_accessible_name",
+              "applied_to": "html/input(type-text)_element",
+              "result": "fail"
```
Comment (mfairchild365) [inline data/tests/tech/aria/aria-placeholder.json:256]: I'm hearing the correct output when I test it myself here too. Might be good to verify all the Edge output.

File: data/tests/tech/aria/aria-placeholder.json [hunk 2]
```diff
@@ -0,0 +1,1693 @@
...
+          "os_version": "10.0.18363",
+          "date": "2021-11-26"
+        }
+      }
+    },
+    "narrator": {
+      "browsers": {
+        "chrome": {
+          "at_version": "10.0.18363",
+          "browser_version": "91",
```
Comment (mfairchild365) [inline data/tests/tech/aria/aria-placeholder.json:1620]: This should be fine, but is it possible test with a newer version?

**Review by mfairchild365 — COMMENTED** [review]
Thank you so much for this! I checked a sample of the results, many were spot on, but a few, especially those around Edge, were a bit different from what I found. I think it would be good to get to the bottom of that. It could be that I'm doing something wrong, but I think the next step is for you to verify your results, especially for edge.

Also, would you be interested in setting up and testing Orca at some point? Not a hard requirement for this PR, but it could be helpful in the future. I'd be happy to help you get that set up.

File: data/tests/tech/aria/aria-placeholder.json (hunk 1 above)
Comment (RavenAlly) [reply data/tests/tech/aria/aria-placeholder.json]: I have retested and updated all Windows results using a different machine.

File: data/tests/tech/aria/aria-placeholder.json [hunk 3]
```diff
@@ -0,0 +1,1693 @@
...
+          "os_version": "10.0.18363",
+          "date": "2021-11-26"
+        },
+        "edge": {
+          "at_version": "10.0.18363",
+          "browser_version": "96",
+          "os_version": "10.0.18363",
```
[the remark this answers is not in the document]
Comment (RavenAlly) [reply data/tests/tech/aria/aria-placeholder.json]: Sorry, I was using System Information version. Have updated to use About version format instead.

File: data/tests/tech/aria/aria-placeholder.json (hunk 2 above)
Comment (RavenAlly) [reply data/tests/tech/aria/aria-placeholder.json]: That's what happens when I test on my work machine - very locked down.
I've retested all Windows results on a personal machine instead, much better with the newer versions all round.

**Comment by mfairchild365** [comment]
Awesome work! Thank you so much!

**Pull request merged**